core/templates/dev/head/pages/community-dashboard-page/services/contribution-and-review.service.spec.ts

```diff
@@ -0,0 +1,114 @@
+// Copyright 2018 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Tests that the contribution-and-review service is working as
+ * expected.
+ */
+
+// TODO(#7222): Remove the following block of unnnecessary imports once
+// UserService.ts is upgraded to Angular 8.
+import { UpgradedServices } from 'services/UpgradedServices';
+// ^^^ This block is to be removed.
+
+require('domain/utilities/url-interpolation.service.ts');
+require(
+  'pages/community-dashboard-page/services/' +
+  'contribution-and-review.service.ts');
+
+describe('ContributionAndReviewService', function() {
+  var ContributionAndReviewService, $httpBackend, $q, UrlInterpolationService;
+  var suggestion1;
+  var opportunityDict1;
+  var mockSuggestionsBackendObject;
+  var expectedSuggestionDict;
+  var suggestionIdToSuggestions;
+  var onSuccess;
+
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(angular.mock.module('oppia', function($provide) {
+    var ugs = new UpgradedServices();
+    for (let [key, value] of Object.entries(ugs.getUpgradedServices())) {
+      $provide.value(key, value);
+    }
+  }));
+
+  beforeEach(angular.mock.inject(function($injector) {
+    ContributionAndReviewService = $injector.get(
+      'ContributionAndReviewService');
+    UrlInterpolationService = $injector.get(
+      'UrlInterpolationService');
+    $httpBackend = $injector.get('$httpBackend');
+    $q = $injector.get('$q');
+
+    suggestion1 = {
+      suggestion_id: 'suggestion_id_1',
+      target_id: 'skill_id_1',
+    };
+    opportunityDict1 = {
+      skill_id: 'skill_id_1',
+      skill_description: 'skill_description_1',
+    };
+    mockSuggestionsBackendObject = {
+      suggestions: [
+        suggestion1
+      ],
+      target_id_to_opportunity_dict: {
+        skill_id_1: opportunityDict1,
+      },
+    };
+    expectedSuggestionDict = {
+      suggestion: suggestion1,
+      details: opportunityDict1,
+    };
+    suggestionIdToSuggestions = {};
+    onSuccess = function(data) {
+      suggestionIdToSuggestions = data;
+    };
+  }));
+
+  afterEach(function() {
+    $httpBackend.flush();
```

Comment:
There should be a check that there are no outstanding requests. The $httpbackend.flush should happen just after where you expect the http request to be fired. Please see https://github.com/oppia/oppia/pull/8180/files#diff-a142d023fe7834e15ce89adb1a398898R73-R76 for an example of how to use $httpbackend.

Reply:
Can you elaborate? In this case, $httpbackend.flush is being called after the ContributionAndReviewService calls where the http requests are being made.

Reply:
What I meant is include the httpbackend.flush() as part of the it block, and in the after each, have the after each block have these statements
``` 
afterEach(function() {
    $httpBackend.verifyNoOutstandingExpectation();
    $httpBackend.verifyNoOutstandingRequest();
  });
```